src/values/AbstractValue.js

```diff
@@ -376,6 +376,25 @@ export default class AbstractValue extends Value {
     throw new FatalError();
   }
 
+  static createFromCoercionOp(
+    realm: Realm,
+    coercionType: string,
+    operand: AbstractValue,
+    resultType: typeof Value,
+    loc?: ?BabelNodeSourceLocation
+  ): AbstractValue {
+    let resultTypes = new TypesDomain(resultType);
+    let resultValues = ValuesDomain.topVal;
+    let hash;
+    let operands;
+    [hash, operands] = hashCall(coercionType, operand);
+    let Constructor = Value.isTypeCompatibleWith(resultType, ObjectValue) ? AbstractObjectValue : AbstractValue;
+    let result = new Constructor(realm, resultTypes, resultValues, hash, operands, ([x]) => t.sequenceExpression([x]));
```

Comment:
There shouldn't be a need for a sequence expression here. Also, it seems more useful to let the caller supply the build function.

Reply:
Hard to say how useful it is to provide the build function since every other case that needs to do this currently uses createFromTemplate to do it.

If this is truly just a "cast" without runtime semantics, then it seems that going through the builder mechanism is unnecessary and we should really just make this a straight alias.

I don't know how to fix this invariant if I don't wrap it in a sequenceExpression.

https://github.com/facebook/prepack/blob/9949c16e45fca7c38a4719867a81c33a20cb6e21/src/serializer/ResidualHeapSerializer.js#L1483-L1486

I'm pretty sure that this check is too aggressive but I'm not familiar enough with the serializer to know how to fix it.

Reply:
I have the following in a local commit that I'm not ready to put up in a pull request:

static createFromBuildFunction(
   realm: Realm,
   resultType: typeof Value,
   args: Array<Value>,
   buildFunction: AbstractValueBuildNodeFunction,
   optionalArgs?: {|
     kind?: string,
     returnValueOf?: ECMAScriptSourceFunctionValue,
   |}
 ): AbstractValue | UndefinedValue {
   let types = new TypesDomain(resultType);
   let values = ValuesDomain.topVal;
   let Constructor = Value.isTypeCompatibleWith(resultType, ObjectValue) ? AbstractObjectValue : AbstractValue;
   let result = new Constructor(realm, types, values, 0, args, buildFunction);
   if (optionalArgs !== undefined) {
     if (optionalArgs.kind !== undefined) result.kind = optionalArgs.kind;
     if (optionalArgs.returnValueOf !== undefined) result.returnValueOf = optionalArgs.returnValueOf;
   }
   return result;
 }

Reply:
What does hash `0` mean?

Reply:
Nothing good. It would be better to incorporate the kind and any arguments into the hash.